### PR TITLE
support tls termination for tcp traffic

### LINKED ISSUE #1391: TLS Termination support for TLS listener
Is TLS Termination for TLS listener for Downstream connections supported with Envoy Gateway as described in this document?
https://gateway-api.sigs.k8s.io/v1alpha2/guides/tls

I mean the second case of the table where:
Listener Protocol = TLS
TLS Mode = Terminate
Route Type =TCPRoute

At the moment when trying to apply the below config (may be doing something wrong) getting the message in the Gateway status:
"**TLS mode "Terminate" is not supported, TLS mode must be Passthrough**"

```
cat <<EOF | kubectl apply -f -
apiVersion: gateway.networking.k8s.io/v1beta1
kind: GatewayClass
metadata:
  name: eg
spec:
  controllerName: gateway.envoyproxy.io/gatewayclass-controller

---
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: eg
  namespace: envoy-gateway-system
spec:
  gatewayClassName: eg
  listeners:
    - name: tls
      protocol: TLS
      port: 8443
      tls:
        mode: Terminate
        certificateRefs:
        - kind: Secret
          name: envoy-gateway
EOF
```

*Note*:
Used helm to install Envoy Gateway:
```
helm install eg oci://docker.io/envoyproxy/gateway-helm --version v0.4.0 -n envoy-gateway-system --create-namespace
```

### PR DESCRIPTION
**Pull request by tanujd11**
**What type of PR is this?**
 Feature to provide TLS termination with TCP traffic

**What this PR does / why we need it**:

Supports the following configuration:
Listener Protocol = TLS
TLS Mode = Terminate
Route Type =TCPRoute

**Which issue(s) this PR fixes**:

Fixes #1391

### CODE REVIEW

**Review by Xunzhuo — APPROVED** [review]
Thanks for working on this : ) LGTM

**Comment by zirain** [comment]
@tanujd11 is there any conformace test for this?

**Comment by tanujd11** [comment]
> @tanujd11 is there any conformace test for this?

No @zirain, I haven't added any e2e test for this yet. Shall I do it in this PR?

**Review by qicz — APPROVED** [review]
Greate Job.

**Review by qicz — APPROVED** [review]
Greate Job.

**Review by qicz — DISMISSED** [review]
Greate Job.

File: internal/ir/xds.go [hunk 1]
```diff
@@ -612,7 +612,9 @@ type TCPListener struct {
 	Port uint32
 	// TLS information required for TLS Passthrough, If provided, incoming
 	// connections' server names are inspected and routed to backends accordingly.
-	TLS *TLSInspectorConfig
+	TLSInspectorConfig *TLSInspectorConfig
```
Comment (arkodg) [inline internal/ir/xds.go:615]: thoughts on making this
```
tls:
  passthrough:
  terminate:
```
?

Comment (tanujd11) [reply internal/ir/xds.go:615]: I feel in future we could use SNIs to route traffic in TCP terminate mode as well? Wdyt? Istio does it on the basis of gateway hostname. However at the moment it does not make sense as TCPRoute does not have any SNI based routing. Maybe I am missing something. Wdyt?

Comment (arkodg) [reply internal/ir/xds.go:615]: that case is handled in the `HTTPListener` https://github.com/envoyproxy/gateway/blob/334422166f208a03440fcec1534c8fe4a2aab3ef/internal/ir/xds.go#L117, the Gateway.Listener.Hostname API field internally performs a SNI match at the xds Listener level

**Review by arkodg — APPROVED** [review]
LGTM thanks for adding this feature !
can you also raise a GH issue here or in upstream to add a conformance test for it, TIA

**Comment by tanujd11** [comment]
https://github.com/kubernetes-sigs/gateway-api/issues/2060 for tracking conformance test related to this

**Review by zirain — APPROVED** [review]

**Pull request merged**